change torch.testing helper asserts to checks

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55385 Add support for checking tensor containers in `torch.testing`
* #55145 make torch.testing error messages more expressive
* #54784 add tests for torch.testing
* #54781 torch.testing allclose -> close
* **#54780 change torch.testing helper asserts to checks**
* #54769 make torch.testing asserts importable

- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

Differential Revision: [D27438849](https://our.internmc.facebook.com/intern/diff/D27438849)

pmeier committed:
change torch.testing helper asserts to checks

[ghstack-poisoned]

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

[ghstack-poisoned]

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

[ghstack-poisoned]

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

[ghstack-poisoned]

pmeier commented:
One problem with this approach is that the type of the suggested error is lost. It might be much cleaner to return `Optional[Exception]` and do 

```python
exc = check_foo()
if exc:
    raise exc
``` 

With this we avoid the double return value and keep the suggested error type.

mattip commented:
> With this we avoid the double return value and keep the suggested error type.

There are two goals here:
- display a less confusing stack trace of exceptions
- ensure the error message contains enough context to work out exactly why the test failed and what needs to be done to fix either the test or the implementation. One way to do this is to have a concise message of which argument and what attribute of that argument caused the exception. 

As long as the exception returned from the internal helper function does the second thing, I think this is fine.

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

[ghstack-poisoned]

mruberry reviewed (approved):
Cool!

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

Differential Revision: [D27438849](https://our.internmc.facebook.com/intern/diff/D27438849)

[ghstack-poisoned]

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

Differential Revision: [D27438849](https://our.internmc.facebook.com/intern/diff/D27438849)

[ghstack-poisoned]

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

Differential Revision: [D27438849](https://our.internmc.facebook.com/intern/diff/D27438849)

[ghstack-poisoned]

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

Differential Revision: [D27438849](https://our.internmc.facebook.com/intern/diff/D27438849)

[ghstack-poisoned]

pmeier committed:
Update on "change torch.testing helper asserts to checks"


- In #53152 we opted to use `tb=native`. Thus, regardless if we use `pytest` to run the tests `__tracebackhide__` is not honored. and additional layers of helper functions make the traceback harder to parse. To overcome this, we change the internal helpers to return `ok: bool, msg: Optional[str]` and only raise the error in the top level function. We do that already in the current implementation that we are trying to replace:
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L92-L93
    https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/__init__.py#L112

Differential Revision: [D27438849](https://our.internmc.facebook.com/intern/diff/D27438849)

[ghstack-poisoned]
